### DISCUSSION TITLE
Test case fails because of StackState's hidden values.

### DISCUSSION THREAD

**Discussion by cihankisakurek** [pointfreeco/swift-composable-architecture#2412] [General]
### Description

Hello, 
I am currently updating my TCA 0.55 project to TCA 1.1.0 and new navigation store. Migration part went pretty well however I am stuck in an issue with test store. 

Test function looks like this.

<img width="1216" alt="Bildschirm­foto 2023-08-25 um 17 05 43" src="https://github.com/pointfreeco/swift-composable-architecture/assets/102791209/5360d505-c3fa-47fd-bbcd-b7cb420c6cd9">

and the error output is 

```
testOpenURLAcceptInvitation(): A state change does not match expectation: …

      AppDomain.State(
        _destination: .tournamentList(
          TournamentListDomain.State(
            _destination: nil,
            path: [
    −         #1: .tournament(…)
    +         #0: .tournament(…)
            ],
            tournaments: [],
            emptyState: nil,
            isLoading: false
          )
        ),
        receivedURL: nil,
        userAuthState: nil,
        pendingInvitationId: "64e8ad3d104b8781570765bf"
      )

(Expected: −, Actual: +)
```

which was very confusing for me and I dug a little bit deeper and in TestStore.swift:1019 there is an equality check that fails and if I print and diff the expectedWhenGivenPreviousState and actual state I got this 

<img width="1055" alt="Bildschirm­foto 2023-08-25 um 17 23 53" src="https://github.com/pointfreeco/swift-composable-architecture/assets/102791209/4dc26477-e425-440f-817e-c6d05d5ca9b5">

which confused me some more.

Also I have a very similar test case that works. 

<img width="1066" alt="Bildschirm­foto 2023-08-25 um 17 29 04" src="https://github.com/pointfreeco/swift-composable-architecture/assets/102791209/b993f62a-d7ce-438b-abd3-f2d88064a4c6">

Only difference I can see is that in the second case I send the action .navigateToTournament and in failing case it is a receive. 

Thanks in advance.

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

TestStore's send and receive calls should behave same.

### Actual behavior

there is a discrepancy between TestStore.send and TestStore.receive

### Steps to reproduce

please see the explanation 

### The Composable Architecture version information

1.1.0

### Destination operating system

iOS 16 

### Xcode version information

Version 14.3 (14E222b)

### Swift Compiler version information

```shell
swift-driver version: 1.75.2 Apple Swift version 5.8 (swiftlang-5.8.0.124.2 clang-1403.0.22.11.100)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @cihankisakurek, this is discussed in our episodes and in the docs, but in order for `StackState` to work the way it does it needs to maintain an internal list of IDs. And if you ever create new `StackState` from scratch (like you do in your test), then you will generate all new IDs.

Instead of doing that there are helpers on `StackState` that allow you to traverse into an element and further into a case of the `Path.State` enum and make a mutation:

```swift
state.path[id: 0, case: /Path.State.feature] = …
```

This will preserve the IDs and allow you to mutate state.

Also, this is not an issue with the library so I am going to convert it to a discussion. Feel free to ask more questions over there.

**Reply by mbrandonw** [reply]
It looks like you also have some `Destination` enum state too, and there is a similar helper on that too, but it uses the `@PresentationState` projected value:

```swift
state.$destination[case: /Destination.State.feature] = …
```

**Reply by cihankisakurek** [reply]
I see. But it doesn't explain to me why it is passing the test in the second case.

**Reply by mbrandonw** [reply]
We make a copy of the `\.stackElementID` generator when invoking the `send` and `receive trailing closures:

https://github.com/pointfreeco/swift-composable-architecture/blob/6dfdd189064f96fb2543265cf55148ec731c70b3/Sources/ComposableArchitecture/TestStore.swift#L900-L905

This means if you are testing a feature that appends a new element to a stack, and so it generates id `#0`, and in the assert you do `path = […]`, then you will generate another id `#0` in that assertion.

Honestly, in retrospect, I do not think we should have done that logic in the test store. And we should probably remove it in 2.0. Instead your second test should be written as:

```swift
state.$destination[case: /Destination.Action.feature]?.path[id: 0] = …
```

This makes it clear you know what generational ID is being used for your newly pushed feature.

**Reply by cihankisakurek** [reply]
Thank you for your explanation it make sense now. Yes I agree.  I always assumed they would behave exactly the same (since a user action and a side effect should be same) but I guess it is a design choice in some sense. 

Anyway thanks for your time and also for this great library!

**Comment by seanmrich** [comment]
I ran into a similar problem while testing path mutations. In my setup, child features use delegate actions to notify a root feature to modify the navigation stack. Here, the `Child` feature is just a button that should push an `Other` feature.

```swift
@Reducer
struct Child {
  enum Action {
    case itemTapped
  }
}

struct ChildView: View {
  let store: StoreOf<Child>
  var body: some View {
    Button("Item") {
      store.send(.itemTapped)
    }
  }
}
```

The `Root` feature listens for that delegate action and pushes `Other` onto the stack.

```swift
@Reducer
struct Root {
  @Reducer(state: .equatable)
  enum Path {
    case child(Child)
    case other(Other)
  }
  struct State: Equatable {
    var path = StackState<Path.State>()
  }
  enum Action {
    case path(StackAction<Path.State, Path.Action>)
  }
  var body: some ReducerOf<Self> {
    Reduce { state, action in
      switch action {
      case .path(.element(_, action: .child(.itemTapped))):
        state.path.append(.other(Other.State()))
        return .none
      case .path:
        return .none
      }
    }
    .forEach(\.path, action: \.path)
  }
}
```

This setup works fine. Here's how I expect to test that the navigation stack is modified correctly:

```swift
@MainActor
func testItemTapped() async {
  let store = TestStore(
    initialState: Root.State(
      path: StackState([
        .child(Child.State())
      ])
    ),
    reducer: { Root() }
  )
  await store.send(.path(.element(id: 0, action: .child(.itemTapped)))) {
    $0.path = StackState([
      .child(Child.State()),
      .other(Other.State())
    ])
  }
}
```

This test fails as described by the OP because the StackElementID's don't line up. If instead of explicitly describing the state, I use mutations, the test passes.

```swift
await store.send(.path(.element(id: 0, action: .child(.itemTapped)))) {
  $0.path.append(.other(Other.State()))
}
```

This isn't a huge issue, but the test feels less satisfying because I'm not really confirming the entire contents of the stack state. Am I executing these tests correctly?

**Reply by mbrandonw** [reply]
> This isn't a huge issue, but the test feels less satisfying because I'm not really confirming the entire contents of the stack state. Am I executing these tests correctly?

You can also assert like this if you want to assert on the full stack:

```swift
await store.send(.path(.element(id: 0, action: .child(.itemTapped)))) {
  $0.path[id: 0] = .other(Other.State())
  $0.path[id: 1] = .other(Other.State())
  XCTAssertEqual($0.path.count, 2)
}
```

And I suppose we could consider a dictionary-like way of asserting on the stack:

```
$0.path = [
  0: .other(Other.State()),
  1: .other(Other.State()),
]
```

We would have to think about it a bit more.

**Reply by seanmrich** [reply]
Thanks for that. Here's a little extension that gets you pretty close:
```swift
extension StackState {
  mutating func assert(_ expected: OrderedDictionary<StackElementID, Element>) {
    for (id, element) in expected {
      guard self[id: id] == nil else { continue }
      self.append(element)
    }
    for removedId in ids.subtracting(expected.keys) {
      self[id: removedId] = nil
    }    
  }
}
```

Usage:
```swift
await store.send(.path(.element(id: 0, action: .child(.itemTapped)))) {
  $0.path.assert([
    0: .child(Child.State()),
    1: .other(Other.State())
  ])
}
```

Edited: Originally only handled appending paths. Modified to also allow for popped elements
